tkg/client/upgrade_cluster.go

```diff
@@ -181,18 +181,35 @@ func (c *TkgClient) UpgradeCluster(options *UpgradeClusterOptions) error {
 	return c.DoLegacyClusterUpgrade(regionalClusterClient, currentClusterClient, options)
 }
 
-func (c *TkgClient) DoLegacyClusterUpgrade(regionalClusterClient, currentClusterClient clusterclient.Client, options *UpgradeClusterOptions) error {
+func (c *TkgClient) DoLegacyClusterUpgrade(regionalClusterClient, currentClusterClient clusterclient.Client, options *UpgradeClusterOptions) (err error) {
 	log.Info("Verifying kubernetes version...")
-	err := c.verifyK8sVersion(currentClusterClient, options.KubernetesVersion)
+	err = c.verifyK8sVersion(currentClusterClient, options.KubernetesVersion)
 	if err != nil {
 		return errors.Wrap(err, "kubernetes version verification failed")
 	}
 
+	// backup tkr version for current cluster
+	currentClusterInfo := regionalClusterClient.GetClusterStatusInfo(options.ClusterName, options.Namespace, currentClusterClient)
+	currentTkrVersion := currentClusterInfo.ClusterObject.Labels["tanzuKubernetesRelease"]
+
+	log.Infof("Upgrading kubernetes cluster from `%v` to `%v` version", currentTkrVersion, options.TkrVersion)
+
 	err = c.addKubernetesReleaseLabel(regionalClusterClient, options)
 	if err != nil {
 		return errors.Wrapf(err, "unable to patch the cluster object with TanzuKubernetesRelease label")
 	}
 
+	// restore tkr version if upgrade failed after tkr label updated
+	defer func() {
+		if err != nil {
+			log.Infof("Restoring kubernetes cluster to tkr `%v` version", currentTkrVersion)
+			options.TkrVersion = currentTkrVersion
+			if restoreErr := c.addKubernetesReleaseLabel(regionalClusterClient, options); restoreErr != nil {
```

Comment:
Hi @hongtaiwang , what if the cluster node is successfully upgraded and all the core addons are upgraded, but one of the addon's reconciliation just got a little timeout? I think switching the tkr version back to origin no matter what the error is will result in some unexpected behavior, like the successfully reconciled core addons will be downgraded , which is either not allowed by kapp or bring compatibility issues.

Reply:
does kapp refer to  `tanzuKubernetesRelease` label for legacy cluster or `topology.version` for cc cluster? 

The revert is only for those fields when upgrade failed. If upgrade failed due to some reason, it's expected that the cluster is broken, and users should be able to re-run the upgrade command to get through it.

Reply:
`tanzuKubernetesRelease` took effect on tanzu cli, addons-manager, pinniped... I am not sure what is the full consequence for these components to set it back and forth.

And also `tanzuKubernetesRelease` is a legacy tkr label. In ClusterClass clusters the tkr label seems to be `run.tanzu.vmware.com/tkr`.

Reply:
tanzu cli upgrade cluster read `tanzuKubernetesRelease` in legacy clusters , while `topology.version`  in cc clusters.

Reply:
From the addon-manager's perspective, this `tanzuKubernetesRelease` decide which core package repository should be deployed to to legacy wl cluster. Once it is updated the core addon's upgrade is started.

Reply:
So what I mean is this `tanzuKubernetesRelease` is a very high level information that multiple components are depends on it. If the cluster upgrade is failed and we want to retry, it is better to have something like a `force upgrade` to adjust the cluster into the desired state, instead of `fake` the cluster that the upgrade did not happened before.

Reply:
And maybe the better solution is to block the upgrade from beginning if it tries to skip a minor version

Reply:
But I still think upgrade tag should be the last step for upgrade cluster since we should not set the cluster to new version if the upgrade process is not finished completely. 

And of course the impact for addon-manager should be taken into consideration. Do you know if addon-manager could work after the following update?
* set tkr v1 -> v2 during upgrade
* upgrade fail after addon-manager updated, set back v2 -> v1 (maybe addon-manager will be broken now)
* re-run upgrade command, v1 -> v2 again ===> could addon-manager work correctly in this case?

Reply:
> upgrade tag should be the last step for upgrade cluster 

This was the previous behavior that brought problems. The core addon's upgrade path is `upgrade tkr tag in cluster-> addons manager upgrade core addon's repo -> kapp-controller upgrade core addon`

In this path if the tkr is updated at the final step, then this means the core addons got updated after node upgrade, but if the old addons were not compatible with new version kubernetes nodes, the upgrade will hang on
* Node upgrade stuck because nodes not ready, due to old core addons not compatible
* Addons cannot be upgraded because nodes upgrade was not finished successfully